### PR TITLE
Outbox Bugfix: EndpointName/MsgId, cooperative deletes

### PR DESCRIPTION
**Pull request by DavidBoike**
Connects to #117

### CODE REVIEW

**Comment by DavidBoike** [comment]
While the build succeeds (and that's all GitHub cares about), Acceptance tests are failing.

* [On MSMQ Transport](http://builds.particular.net/viewLog.html?buildId=128487&tab=buildResultsDiv&buildTypeId=NServiceBusNHibernate_2AcceptanceTests)
  * MSSQL: 2 failing tests, reproduced 1 locally, unique constraint because of shared saga table. Cleared the database, reran, passes. I consider these ignorable.
  * Oracle: 2 failing tests
* [On SQL Server Transport](http://builds.particular.net/viewLog.html?buildId=128485&tab=buildResultsDiv&buildTypeId=NServiceBusNHibernate_AcceptanceTestsSqlServer)
  * MSSQL: 2 failing tests
  * Oracle: 16 failures

Do we have any docs on how to set up Oracle for running acceptance tests locally, or for how to run tests locally targeting the SQL transport?

File: src/NServiceBus.NHibernate/Outbox/OutboxPersister.cs [hunk 1]
```diff
@@ -96,13 +105,11 @@ public void RemoveEntriesOlderThan(DateTime dateTime)
...
-                            .List();
+                        var queryString = string.Format("delete from {0} where Dispatched = true And DispatchedAt < :date", typeof(OutboxRecord));
 
-                        foreach (var record in result)
-                        {
-                            session.Delete(record);
-                        }
+                        session.CreateQuery(queryString)
+                            .SetDateTime("date", dateTime)
+                            .ExecuteUpdate();
```
Comment (colin-higgins) [inline src/NServiceBus.NHibernate/Outbox/OutboxPersister.cs]: Nice improvement here. Why make the round trip.

**Comment by DavidBoike** [comment]
@Particular/engineers I need help to know how to proceed with these acceptance tests.

**Comment by andreasohlund** [comment]
@johnsimons would you be able to help @DavidBoike with the oracle failures?

**Comment by johnsimons** [comment]
Sure, what are the failures ?

**Comment by andreasohlund** [comment]
@johnsimons see https://github.com/Particular/NServiceBus.NHibernate/pull/122#issuecomment-147850895

**Comment by DavidBoike** [comment]
For SqlServerTransport + MSSQL Persistence:

* When_reply_from_a_finder.Should_be_received_by_handler - This is a dirty sharing problem, and succeeds with a clean database.
* When_individualization_is_enabled.Should_use_the_configured_differentiator - This appears to have _nothing_ to do with NHibernate. The created address is `IndividualizationIsEnabled.IndividualizedEndpoint` and it was expecting the string `-something` as a discriminator because the endpoint calls:

    EndpointSetup<DefaultServer>(c=>c.ScaleOut().UniqueQueuePerEndpointInstance(discriminator));

I'm inclined to ignore both in order to get the bugfix released.

**Comment by DavidBoike** [comment]
I was able to get the Oracle tests running locally by hacking in the [Oracle.ManagedDataAccess NuGet Package](https://www.nuget.org/packages/Oracle.ManagedDataAccess). Long term I think we should consider switching to that (no external dependencies!) but that's for another PR.

So on Oracle...

* Same dirty sharing problem with When_reply_from_a_finder.Should_be_received_by_handler. Succeeded on clean DB.
* Same SQL Transport problem with When_individualization_is_enabled.Should_use_the_configured_differentiator.
* Erratic problems on When_sending_to_slr.Should_raise_FinishedMessageProcessing_event. Realized it's a retry test that didn't call `.AllowExceptions()`. But the weird thing is this succeeded on MSSQL. There's something funky going on there because 3 tests share the same file/setup. I did 2 separate commits to add the line, although those would get blown away if the AT source package was updated.
* All other tests succeed locally.

So acceptance tests are a mess, but the SQL Transport related discriminator remains the only thing I can't get to pass.

For a bugfix release, I don't think that's a reason to stop the line. Thoughts? @andreasohlund @johnsimons

**Comment by DavidBoike** [comment]
[Msmq](http://builds.particular.net/viewLog.html?buildId=129231&tab=buildResultsDiv&buildTypeId=NServiceBusNHibernate_2AcceptanceTests) and [SqlTransport](http://builds.particular.net/viewLog.html?buildId=129229&tab=buildResultsDiv&buildTypeId=NServiceBusNHibernate_AcceptanceTestsSqlServer) results. Boggles me that there are so many more failures on SqlTransport on the build server I can't recreate locally.

**Comment by johnsimons** [comment]
@DavidBoike in my experience if ATs are failing that usually means bugs, so not sure it is advisable to proceed.
Regarding Oracle, I can show u my setup if needed ?

**Comment by DavidBoike** [comment]
What about failing on build server vs. locally?

Is my hack to use the managed, no-dependency Oracle client valid? Without that, I was getting "Oracle.DataAccess.Client.OracleException : The provider is not compatible with the version of Oracle client"

**Comment by johnsimons** [comment]
@DavidBoike here is how I installed Oracle https://github.com/Particular/NServiceBus.NHibernate/issues/5#issuecomment-36964925

**Comment by johnsimons** [comment]
I can run the AT if you want me to?
Are you running NH against SqlTransport ?

**Comment by DavidBoike** [comment]
@johnsimons Yes please! I figured out how to run it in all 4 combinations, at least with the help of the managed Oracle client. My results are all noted above.

**Comment by johnsimons** [comment]
@DavidBoike isn't this a hotfix, hence be out of master ?

**Comment by johnsimons** [comment]
What is the performance impact of this change ?

**Comment by johnsimons** [comment]
We should update the NServiceBus.AcceptanceTesting and NServiceBus.AcceptanceTesting.Sources to 5.2.8

**Comment by DavidBoike** [comment]
We went all over the place defining this issue. When it was going to be a schema change I would definitely have said hotfix. Since we aren't, and are just changing behavior and no API or schema, I could go either way.

Except if we release a minor then we still need to support the old broken minor for a year. Guess that would seal it for me: should be a patch release.

For perf, the lookup was a PK lookup and is now an OR lookup on primary key. Still clustered index seeks either way. For the cleanup it is MUCH more efficient doing a direct DELETE WHERE instead of a SELECT + N deletes.

**Comment by johnsimons** [comment]
@DavidBoike not sure if it makes sense to test Sql trasport with Oracle persistence, I don't see anyone using this kind of setup, do you ?

**Comment by DavidBoike** [comment]
No, doesn't really make any sense. It's just what is configured to run right now.

**Comment by andreasohlund** [comment]
The oracle managed driver seems to be broken for DTC https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/particularsoftware/hAn94zv9c-o/DzqxBoFbBQAJ :(

**Comment by DavidBoike** [comment]
These changes went out as hotfixes already. I just lost track of this one. Any reason this can't be merged?

**Comment by DavidBoike** [comment]
This PR started with the intention of releasing Outbox fix as a new minor, but this turned into hotfixes. The important commits were cherry-picked for those hotfixes, and concepts from the others (like custom connection strings for ATs) have made their way into Core.

Closing this PR in favor of simply merging master back into develop.

**Pull request closed without merging**